torch/jit/_recursive.py

```diff
@@ -1,112 +1,399 @@
+        for known_concrete_type, type_ in known_types:
+            if concrete_type.equals(known_concrete_type):
+                return type_
+        return None
 
-    # Copy overloads
-    script_module.__dict__["_overloads"] = dict(getattr(original, "__overloads__", {}))
+    def create_new_type(self, original_type, concrete_type):
+        qualified_name = torch.jit._qualified_name(original_type)
+        type_ = torch._C._make_jit_type_from_concrete_type(qualified_name, concrete_type)
+        if original_type not in self.type_map:
+            self.type_map[original_type] = []
 
-    # Copy links to Python methods so they can be resolved when compiling
-    for name in dir(original):
-        item = getattr(original, name)
-        if hasattr(script_module, name):
-            # Skip Python builtins and all the module methods that are already
-            # attached to this since it inherits from nn.Module
+        self.type_map[original_type].append((concrete_type, type_))
+        return type_
+
+concrete_type_store = ConcreteTypeStore()
+
+def get_type(original):
+    assert isinstance(original, Module)
+    if isinstance(original, torch.jit.RecursiveScriptModule) and not original._initializing:
+        return original._concrete_type, original._c._type()
+
+    concrete_type = get_concrete_type(original)
+    maybe_type = concrete_type_store.find_type(type(original), concrete_type)
+    if maybe_type is not None:
+        type_ = maybe_type
+    else:
+        type_ = concrete_type_store.create_new_type(type(original), concrete_type)
+
+    return concrete_type, type_
+
+def create_methods_from_stubs(cpp_mod, concrete_type, stubs):
+    defs = [m.def_ for m in stubs]
+    rcbs = [m.resolution_callback for m in stubs]
+    defaults = [get_default_args(m.original_method) for m in stubs]
+    cpp_mod._create_methods(concrete_type, defs, rcbs, defaults)
+
+def compile_unbound_method(cpp_mod, concrete_type, fn):
+    if _jit_internal.is_ignored_fn(fn):
+        return None
+    stub = make_stub(fn)
+    with torch.jit._disable_emit_hooks():
+        # We don't want to call the hooks here since the graph that is calling
+        # this function is not yet complete
+        create_methods_from_stubs(cpp_mod, concrete_type, (stub,))
+    return stub
+
+# TODO redoc This is called in the following ways:
+# 1. From recursive script. py_module is an nn.Module
+# 2. From ScriptMeta2. py_module is a ScriptModule
+# 3. From tracing, in the construction of a traced module. py_module is a
+#    ScriptModule with a fresh ._c already set
+# 4. From tracing, if we have exports set on the traced module. Then its called
+#    directly to just compile the exports.
+def create_script_module(original_module, stubs, fresh_type=False):
+    assert isinstance(original_module, torch.nn.Module)
+    check_module_initialized(original_module)
+
+    concrete_type, module_type = get_type(original_module)
+    if fresh_type:
```

Comment:
This seems weird. There seem to be a lot of functions on cpp_mod that should really be on ClassType instead.

Reply:
This is where we actually set the data, hence why it is on the module object and not the type. The `register_foo` is a little confusing—it's actually a set, not a register, but making it `set` instead of `register` necessitated a little too much extraneous refactoring (we might be able to change it after @ZolotukhinM's PR)